pkgs/build-support/setup-hooks/make-wrapper.sh

```diff
@@ -23,6 +23,8 @@ assertExecutable() {
 
 # --prefix          ENV SEP VAL   : suffix/prefix ENV with VAL, separated by SEP
 # --suffix
+# --prefix-dedup    ENV SEP VAL   : same as --prefix/--suffix, but adding the same value twice skips it
+# --suffix-dedup
```

Comment:
Are you sure `--suffix-dedup` should *move* the values to the end? What is the use case for that?

My instinct is that `--prefix` should always remove duplicates, and `--suffix` should do nothing for the values already present. I would only add the extra `-dedup` options if the performance hit is a concern.

(But take my opinions with a grain of salt, I am only here occasionally.)

Reply:
Maybe you should look at the comments in https://github.com/NixOS/nixpkgs/pull/127678

Reply:
It should always remove dupes because we need this everywhere and are not going to change every occurrence.

Reply:
I've made all the `prefix-*` and `suffix-*` actions deduplicate values now and removed the `*-dedup` actions. I now also deduplicate the components of the value.

Reply:
After looking at this again, I really think that `--suffix` should be like `addToSearchPath`. The model is: the position of the first occurrence of an entry has a meaning (it takes precedence over the following entries), the position of the later occurrences has no meaning (except for php, but that's unintended behaviour). So it's safe to remove late occurrences, but it's not safe to remove the first one.

Hence `--prefix` should tack the entry to the left, and remove the duplicates, if any; `--suffix` should append the entry only if it's not already present.

Unfortunately that means that `--prefix` and `--suffix` must generate different code (sorry @talyz! more work I guess).

PS: and `--prefix` should only remove one entry. If there are actual duplicates, they won't come from `make-wrapper.sh`, and it's better to leave them unaltered – it means there is a bug somewhere else.

Reply:
Well, if I understand you correctly, you're essentially saying that we shouldn't deduplicate at all - why not? There's a clear need for deduplication, as I see it, and I don't know what the harm in doing so would be, even when not strictly necessary.

Reply:
I am suggesting something slightly different, which is, just ensure that the wrapper never introduces a *new* duplicate. So in the end there won't be duplicates at all (this is how `addToSearchPath` works).

Reply:
Okay. I've implemented the happy medium - full deduplication when prefixing and going the `addToSearchPath` route when suffixing. Documentation and naming still needs to be updated, but I want to check that you're okay with this before putting in more work.

Reply:
Thanks! `suffix` looks right to me. It could be pulled out of `dedupAdd` now, but we can talk style once it's settled.

Re `prefix`, I am still doubtful on full deduplication. I don't like that every wrapper in nixpkgs has to walk through every entry, it's a big tradeoff. Wrappers should do exactly what they need, and nothing more.

Are you sure that my suggested strategy does not work (i.e. `prefix` checks & removes only the first instance, which can be done without a loop)? Is there a scenario I am missing?

Reply:
The plan was to change the name of `dedupAdd` to `addValue` or something similar.

I really think you're overestimating the impact of this loop. In any realistic scenario, its impact will be in the low millisecond range.

I'm not opposed to it if you can find a good way to perform it without sacrificing reliability and readability.

Reply:
The regex and `BASH_REMATCH` are a bit ugly, but for instance
```bash
if [[ $PATH =~ ^((.*):)?"$newPath"(:(.*))?$ ]] ; then
 export PATH="$newPath${BASH_REMATCH[2]:+:${BASH_REMATCH[2]}}${BASH_REMATCH[4]:+:${BASH_REMATCH[4]}}"
else
  export PATH="$newPath${PATH:+:$PATH}"
fi
```
does the job without loops. Let's start with: does that fix the issue you want to address?

Reply:
Yeah, with proper regexes all my issues are solvable. I've cleaned this up, making it as readable as I could, and fit it into the current solution.

Reply:
Documented the whole thing and replaced the suffix case with the same match to make it easier to read. I think this done now, except for squashing and rewriting the commit messages.

Reply:
Thanks @talyz for bearing with me!

Reply:
No worries! Thanks for all the pointers! I wasn't aware of `BASH_REMATCH` - really useful stuff! :)

Reply:
Just realised that there *is* a reasonable way without regular expressions (and POSIX for what it's worth).
```bash
export PATH="$newPath${PATH:+:${PATH%%:$newPath}}"
[[ "$PATH" == *:"$newPath":* ]] && export PATH="${PATH%%:$newPath:*}:${PATH#*:$newPath:}"
```
This may be more readable. It may also remove two instances of `$newPath`, but that should be fine. Similar cost? I have absolutely no idea.

...opinions?

Reply:
To me, that's even harder to read, albeit shorter. I think we can stick with the current solution - POSIX compatibility doesn't matter, since we're always using bash and the performance difference will likely be close to immeasurable in most cases.

Reply:
Yeah, that was not great. What about generating something like
```bash
PATH="${PATH:+:$PATH:}"        # add extra separators on both sides if nonempty
PATH="${PATH/:"/nix/foo":/:}"  # remove first (?) /nix/foo
PATH="${PATH/:"/nix/bar":/:}"  # remove /nix/bar
PATH="${PATH/:"/nix/baz":/:}"  # remove /nix/baz
export PATH="/nix/foo:/nix/bar:/nix/baz${PATH%:}" # prepend all values, drop extra trailing separator
```
?

Reply:
Sure, I've implemented a solution based on that idea now. I'm not sure if it's more readable or not, tbh. It probably depends on whether you're more comfortable with globs and variable expansion hacks or regex :shrug: